### ISSUE TITLE
Seg fault in parallel matrix addition

### ISSUE THREAD

**Issue by ben-s-southworth** [mfem/mfem#1238]
In certain cases, I am getting seg faults using any of the matrix addition functions (Add(), ParAdd(), or .Add()). Below is a minimal working example. I think it has to do with multiplying D*D^T, where D is rectangular, to get a smaller square matrix. When I try to add this matrix with a matrix naturally defined on the smaller space, it works for 1 processor, but seg faults for more than one. Unclear if this is on the hypre side or MFEM side, but valgrind traces it as follows:
```
==55251== Invalid read of size 4
==55251==    at 0x89B0D0: hypre_CSRMatrixAddHost (csr_matop.c:70)
==55251==    by 0x47C970: mfem::internal::hypre_ParCSRMatrixAdd(hypre_ParCSRMatrix_struct*, hypre_ParCSRMatrix_struct*) (hypre_parcsr.cpp:1472)
==55250== Invalid read of size 4
==55251==    by 0x470752: mfem::ParAdd(mfem::HypreParMatrix const*, mfem::HypreParMatrix const*) (hypre.cpp:1573)
==55250==    at 0x89B0D0: hypre_CSRMatrixAddHost (csr_matop.c:70)
==55251==    by 0x409EBC: main (test.cpp:52)
==55251==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
```
For 2 processors, I printed both matrices to file and the rows and data all line up and make sense. MWE:
```c++
#include "mfem.hpp"
#include <fstream>
#include <iostream>

using namespace std;
using namespace mfem;

int main(int argc, char *argv[])
{
    // 1. Initialize MPI.
    int num_procs, myid;
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
    MPI_Comm_rank(MPI_COMM_WORLD, &myid);
    const char *mesh_file = "./data/star.mesh";
    int order = 1;

    // Get parmesh
    Mesh *mesh = new Mesh(mesh_file, 1, 1);
    int dim = mesh->Dimension();
    int ref_levels = 1;
    for (int l = 0; l < ref_levels; l++) {
        mesh->UniformRefinement();
    }
    ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
    delete mesh;

    // Finite element spaces
    L2_FECollection *l2_coll = new L2_FECollection(order, dim);
    ParFiniteElementSpace *dg_space = new ParFiniteElementSpace(pmesh, l2_coll); 
    H1_FECollection *h1_coll = new H1_FECollection(order+1, dim);
    ParFiniteElementSpace *h1_space = new ParFiniteElementSpace(pmesh, h1_coll, dim); // dim copies of H1

    // Hypre matrices
    ParBilinearForm* mass_dg = new ParBilinearForm(dg_space); 
    mass_dg->AddDomainIntegrator(new MassIntegrator()); 
    mass_dg->Assemble(); 
    mass_dg->Finalize(); 
    HypreParMatrix* Ma = mass_dg->ParallelAssemble();       

    ParMixedBilinearForm* div = new ParMixedBilinearForm(h1_space, dg_space);
    div->AddDomainIntegrator(new VectorDivergenceIntegrator); 
    div->Assemble(); 
    div->Finalize(); 
    HypreParMatrix* D = div->ParallelAssemble(); 
    HypreParMatrix* Dt = D->Transpose(); 

    // get schur complement S = Ma - D*G
    HypreParMatrix* DG = ParMult(D, Dt); 
    // TODO : BELOW BOTH OF THE ADD FUNCTIONS SEG FAULT WITH MORE THAN ONE PROCESSOR 
    // HypreParMatrix *S = Add(1., *Ma, -1., *DG);
    HypreParMatrix *S = ParAdd(Ma, DG);
 
    // clean up pointers...
}
```

**Comment by mlstowell** [comment]
Hi, @ben-s-southworth ,

Which version of HYPRE are you using?  The two versions I looked at don't have anything interesting happening at line 70 of `csr_matop.c`.

Best wishes,
Mark

**Comment by jandrej** [comment]
The function was introduced by hypre in version 2.18.0.

**Comment by waltia39** [comment]
Security

**Comment by waltia39** [comment]
SECURITY.md

**Comment by tzanio** [comment]
Thank you for reporting this. I will open a new issue to track all the problems that need to be addressed with the newer hypre releases.

**Comment by ben-s-southworth** [comment]
@mlstowell @jandrej I updated to mfem master and hypre master and have the same issue. Valgrind lines are a bit off, but the seg-fault is the same:

```
==66561== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==66561==    at 0x60CE377: raise (raise.c:55)
==66561==    by 0x60CE3EF: ??? (in /usr/lib64/libc-2.17.so)
==66561==    by 0x946A7B: hypre_CSRMatrixAddHost (csr_matop.c:65)
==66561==    by 0x946A7B: hypre_CSRMatrixAdd (csr_matop.c:138)
==66561==    by 0x47F5BB: mfem::internal::hypre_ParCSRMatrixAdd(hypre_ParCSRMatrix_struct*, hypre_ParCSRMatrix_struct*) (hypre_parcsr.cpp:1546)
==66561==    by 0x472B68: mfem::Add(double, mfem::HypreParMatrix const&, double, mfem::HypreParMatrix const&) (hypre.cpp:1567)
==66561==    by 0x40A217: main (paradd.cpp:62)
```

**Comment by liruipeng** [comment]
@ben-s-southworth 

In line 1591 of hypre.cpp `HypreParMatrix * ParAdd(const HypreParMatrix *A, const HypreParMatrix *B)`

I did the following change to use the hypre's implementation of ParcsrAdd and the problem seemed gone

```
   //hypre_ParCSRMatrix * C = internal::hypre_ParCSRMatrixAdd(*A,*B);
   hypre_ParCSRMatrix *C;

   hypre_ParcsrAdd(1.0, *A, 1.0, *B, &C);
```
Mfem guys: Let me know how you would like to proceed. Thanks! 

-Ruipeng

**Comment by v-dobrev** [comment]
Thanks @liruipeng! Do you know which version of hypre introduced the function `hypre_ParcsrAdd`? It'll be best to add a switch for older versions.

**Comment by liruipeng** [comment]
@v-dobrev I think it was 2.14.0
